### PR TITLE
cleaned up version of test coverage script for travis

### PR DESCRIPTION
**Pull request by evertlammerts**
* .travis/test-coverage.sh: I've used this script as the basis for code coverage reports in a little play project of mine (http://github.com/evertlammerts/mcppg). The original script checked the current directory (pwd) separately from all subdirectories, was undocumented and in general just a little messy (`cat` followed by `grep`). I've cleaned it up a bit for my own use.

* .travis.yml: `diff <(gofmt -d .) <("")` results in a `command not found` on travis-ci, without it breaking the build. I don't know why it doesn't break the build, but `test` works nicer than `diff` for testing whether a string is empty.

### CODE REVIEW

**Comment by jonlawlor** [comment]
That does look cleaner, once it is reviewed would you be up for incorporating in the other gonum repositories?  All of them have the same coverage code.

**Comment by jonlawlor** [comment]
Also please open a pull adding yourself to the AUTHORS and CONTRIBUTORS on https://github.com/gonum/license

**Comment by evertlammerts** [comment]
Sure, I'll script something for that. Re AUTHORS / CONTRIB: will do.

File: .travis/test-coverage.sh [hunk 1]
```diff
@@ -1,43 +1,35 @@
...
-	if [ -f profile.out ]
-	then
-		cat profile.out | grep -v "mode: set" >> acc.out
-	fi
-else
-	exit 1
-fi
+testCover() {
+	# set the return value to 1 (succesful)
+	retval=0
```
Comment (jonlawlor) [inline .travis/test-coverage.sh]: Comment is inconsistent with value, should be `# set the return value to 0 (succesful)` ?

File: .travis/test-coverage.sh [hunk 2]
```diff
@@ -1,43 +1,35 @@
...
+	# get the directory to check from the parameter. Default to '.'
+	d=${1:-.}
+	# skip if there are no Go files here
+	ls $d/*.go &> /dev/null || return $retval
+	# switch to the directory to check
+        pushd $d > /dev/null
```
Comment (jonlawlor) [inline .travis/test-coverage.sh]: Should be a tab instead of spaces.

**Comment by jonlawlor** [comment]
LGTM

**Comment by jonlawlor** [comment]
Alright, the logs look good on the servers, would you mind doing the same in other gonum repos that have test_coverage.sh?

**Comment by evertlammerts** [comment]
Working on it

**Comment by evertlammerts** [comment]
Never automated PR's before :-) Hope these work:

=============== lapack
https://github.com/gonum/lapack/pull/207
=============== graph
https://github.com/gonum/graph/pull/185
=============== blas
https://github.com/gonum/blas/pull/193
=============== plot
https://github.com/gonum/plot/pull/328
=============== matrix
https://github.com/gonum/matrix/pull/406
=============== stat
https://github.com/gonum/stat/pull/144
=============== mathext
https://github.com/gonum/mathext/pull/25
=============== diff
https://github.com/gonum/diff/pull/46
=============== floats
https://github.com/gonum/floats/pull/60
=============== unit
https://github.com/gonum/unit/pull/34
=============== internal
https://github.com/gonum/internal/pull/39
=============== optimize
https://github.com/gonum/optimize/pull/190

**Comment by evertlammerts** [comment]
gonum/lapack#207 also needed a small manual adjustment of the gofmt call in .travis.yml because it explicitly globs on go files (`gofmt -d *.go`), which the other repo's don't do. Also, the test-coverage.sh scripts in the blas and matrix repo's included some `\r\n` (0x0D 0x0A) newlines.

**Comment by kortschak** [comment]
>Also, the test-coverage.sh scripts in the blas and matrix repo's included some `\r\n` (0x0D 0x0A) newlines.

Ugh. I hope they are cleaned out now.

BTW Would you share the script you used to automate these. It will be very helpful for Go versions bumps and so on.

File: .travis.yml [hunk 3]
```diff
@@ -17,5 +17,5 @@ script:
  - go get -d -t -v ./...
  - go build -v ./...
  - go test -v ./...
- - diff <(gofmt -d .) <("")
+ - test -z "`gofmt -d .`"
```
Comment (kortschak) [inline .travis.yml]: Can we change this to `test -z "$(gofmt -d .)"` to avoid backticks please.

File: .travis/test-coverage.sh [hunk 4]
```diff
@@ -1,43 +1,35 @@
 #!/bin/bash
 
-# based on http://stackoverflow.com/questions/21126011/is-it-possible-to-post-coverage-for-multiple-packages-to-coveralls
-# with script found at https://github.com/gopns/gopns/blob/master/test-coverage.sh
+PROFILE_OUT=`pwd`/profile.out
```
Comment (kortschak) [inline .travis/test-coverage.sh]: Oops. These too.

Comment (kortschak) [inline .travis/test-coverage.sh]: Or just $PWD.

**Comment by evertlammerts** [comment]
Changed backticks to $() and pwd into $PWD for all repo's:

https://github.com/gonum/integrate/pull/34
https://github.com/gonum/lapack/pull/208
https://github.com/gonum/graph/pull/186
https://github.com/gonum/blas/pull/194
https://github.com/gonum/plot/pull/329
https://github.com/gonum/matrix/pull/407
https://github.com/gonum/stat/pull/145
https://github.com/gonum/mathext/pull/26
https://github.com/gonum/diff/pull/47
https://github.com/gonum/floats/pull/62
https://github.com/gonum/unit/pull/35
https://github.com/gonum/internal/pull/40
https://github.com/gonum/optimize/pull/191

**Comment by evertlammerts** [comment]
I've used [Hub](https://hub.github.com/) to automate this. The script I wrote follows the [contributor workflow](https://hub.github.com/#contributor) on Hub's homepage pretty literally (see the code below): clone -> [make changes] -> fork -> commit -> push -> create-pr.

The script takes repository names as arguments, ie `update.sh integrate blas lapack` to run the script for three repo's. To make it more generic you should pass some parameters... Something like:
* -execute [exec]: executable that makes the actual changes
* -commitmsg [msg]: commit message
* -remote [remote]: name of the remote used for the fork, ie github username
* -prmsg [msg]: PR message

```bash
#!/bin/bash

for repo in "$@"; do
	# clone the repo
	hub clone https://github.com/gonum/${repo}.git > /dev/null

	echo -n "===================="
	pushd $repo

	# make needed changes
	[ -f .travis.yml ] && sed -i.bu 's/`\(gofmt -d [^`]*\)`/$(\1)/g' .travis.yml
	[ -f .travis/test-coverage.sh ] && sed -i.bu 's/`pwd`/$PWD/g' .travis/test-coverage.sh

	# only create the PR if changes were made
	if test -n "$(hub diff)"; then
		# commit, fork, push, pr
		hub commit -am 'removed backticks and replaced pwd'  > /dev/null
		hub fork  > /dev/null
		hub push evertlammerts master  > /dev/null
		hub pull-request -m 'changes based on comments in gonum/integrate#33'
	fi

	popd > /dev/null
done
```

**Comment by evertlammerts** [comment]
Going ever so slightly overboard on the automatic pr stuff: https://gist.github.com/evertlammerts/733b6282fc59b95ca1ae5e107c70b0a5

**Comment by jonlawlor** [comment]
So something like
```
bash update.sh -g gonum -e "/usr/bin/sed -i 's/1\.5\.4/1\.8\.0/g' .travis.yml" -r name -c 'comm msg' -p 'pr msg' lapack plot graph ...
```
?

**Comment by evertlammerts** [comment]
I always get into trouble escaping stuff when I use find -e so the previous version explicitly wanted an executable, not a command. I've changed it to behave as find. But your sed command won't work on my OSX, it complains about quotes. If you remove them, it works:

```
bash update.sh -g gonum -e "/usr/bin/sed -i.bu s/1\.5\.4/1\.8\.0/g .travis.yml" -r name -c 'comm msg' -p 'pr msg' lapack plot graph ...
```

(Sed on OSX needs a value after -i, hence the -i.bu)

I've tested the script up to but not including the creation of a PR. Note that Hub expects your git credentials to be known.

This is just me playing a little bit. The script makes an effort to detect errors but you'll probably run into some uncaught stuff anyway. You can always just use Hub without it.

**Comment by evertlammerts** [comment]
```
$ bash update.sh --help

update -g NAME -e EXEC -r REMOTE -c CMSG -p PRMSG REPO [...]

Required:
-g|--githubuser NAME: name of the user or organisation owning the repo's
-e|--execute CMD:     command to execute, will be run from each repo's working dir
-r|--remote REMOTE:   name of the remote used for the fork, ie github username
-c|--commitmsg CMSG:  commit message
-p|--prmsg PRMSG:     PR message

Optional:
-h|--help:            display this help and exit
```

**Pull request merged**